k8s/reconcilers/deployment.go

```diff
@@ -0,0 +1,679 @@
+		apiClient              utils.ApiClient
+		reconciliationInterval time.Duration
+		pollInterval           time.Duration
+		deleteTimeOut          time.Duration
+		applyTimeOut           time.Duration // TODO: Use reconciliation policy to configure
+		deleteSyncDelay        time.Duration // TODO: Use operator reconcile loop instead of this delay
+		delayFunc              func(time.Duration)
+		deploymentKeyResolver  func(Reconcilable) string
+		deploymentErrorBuilder func(*apimodel.SummaryResult, error, *apimodel.ErrorType)
+		deploymentBuilder      func(ctx context.Context, object Reconcilable) (*apimodel.DeploymentSpec, error)
+	}
+	DeploymentReconcilerOptions func(*DeploymentReconciler)
+	ReconcilerSubject           string
+)
+
+const (
+	defaultTimeout                = 15 * time.Minute
+	defaultReconciliationInterval = 30 * time.Minute
+	defaultPollInterval           = 10 * time.Second
+)
+
+var (
+	_             Reconciler = &DeploymentReconciler{}
+	termialErrors            = map[v1alpha2.State]struct{}{
+		v1alpha2.TimedOut:               {},
+		v1alpha2.TargetPropertyNotFound: {},
+	}
+)
+
+func NewDeploymentReconciler(opts ...DeploymentReconcilerOptions) (*DeploymentReconciler, error) {
+	r := &DeploymentReconciler{
+		deploymentKeyResolver:  defaultDeploymentKeyResolver,
+		deploymentErrorBuilder: defaultProvisioningErrorBuilder,
+		delayFunc:              time.Sleep,
+		applyTimeOut:           defaultTimeout,
+		reconciliationInterval: defaultReconciliationInterval,
+		pollInterval:           defaultPollInterval,
+		deleteTimeOut:          defaultTimeout,
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	if r.finalizerName == "" {
+		return nil, fmt.Errorf("finalizer name cannot be empty")
+	}
+	if r.kubeClient == nil {
+		return nil, fmt.Errorf("kube client cannot be nil")
+	}
+	if r.apiClient == nil {
+		return nil, fmt.Errorf("api client cannot be nil")
+	}
+	if r.deploymentBuilder == nil {
+		return nil, fmt.Errorf("deployment builder cannot be nil")
+	}
+	return r, nil
+}
+
+func (r *DeploymentReconciler) deriveReconcileInterval(log logr.Logger, target Reconcilable) (reconciliationInterval, timeout time.Duration) {
+	rp := target.GetReconciliationPolicy()
+	reconciliationInterval = r.reconciliationInterval
+	timeout = r.applyTimeOut
+	if rp != nil && rp.Type == k8smodel.ReconciliationPolicy_Periodic {
+		if rp.Interval != nil {
+			interval, err := time.ParseDuration(*rp.Interval)
+			if err != nil {
+				log.Info(fmt.Sprintf("failed to parse reconciliation interval %s, using default %s", *rp.Interval, reconciliationInterval))
+				return
+			}
+			reconciliationInterval = interval
+		}
+	}
+	return
+}
+
+// attemptUpdate attempts to update the instance
+func (r *DeploymentReconciler) AttemptUpdate(ctx context.Context, object Reconcilable, log logr.Logger, operationStartTimeKey string) (metrics.OperationStatus, reconcile.Result, error) {
+	if !controllerutil.ContainsFinalizer(object, r.finalizerName) {
```

Comment:
Do we need to check finalizer in each update?

Reply:
I think it should be yes. We need to check in every reconciliation loop. This behavior also exists in current instance_controller.